### PR TITLE
Add responded field to flow runs

### PR DESCRIPTION
**Pull request by rowanseymour**
If contact sends a message as part of a flow run, that flow run is marked as responded. This field is also added to the two main flow run db indexes used by the API, so that APIv2 can support filtering by responded=true.

### CODE REVIEW

**Comment by norkans7** [comment]
Looks good

File: temba/flows/migrations/0046_flowrun_indexes_2.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+IF NOT EXISTS (
+    SELECT 1 FROM pg_class c JOIN pg_namespace n ON n.oid = c.relnamespace
+    WHERE c.relname = 'flows_flowrun_org_id_modified_on_responded' AND n.nspname = 'public') THEN
+    CREATE INDEX flows_flowrun_org_id_modified_on_responded ON flows_flowrun (org_id, modified_on DESC, responded);
+END IF;
+
+IF NOT EXISTS (
+    SELECT 1 FROM pg_class c JOIN pg_namespace n ON n.oid = c.relnamespace
+    WHERE c.relname = 'flows_flowrun_flow_id_modified_on_responded' AND n.nspname = 'public') THEN
+    CREATE INDEX flows_flowrun_flow_id_modified_on_responded ON flows_flowrun (flow_id, modified_on DESC, responded);
```
Comment (nicpottier) [inline temba/flows/migrations/0046_flowrun_indexes_2.py]: Are these perfect indexes when querying against them? I think modified_on needs to be last for the ordering to be correct for those queries

File: temba/flows/migrations/0044_flowrun_responded.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('flows', '0043_populate_exit_type'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='flowrun',
+            name='responded',
+            field=models.BooleanField(default=False, help_text='Whether contact has responded in this run'),
```
Comment (nicpottier) [inline temba/flows/migrations/0044_flowrun_responded.py]: This needs to be nullable no?

File: temba/flows/models.py [hunk 3]
```diff
@@ -3577,6 +3579,11 @@ def add_message(self, msg):
             msg.msg_type = FLOW
             msg.save(update_fields=['msg_type'])
 
+        # if message is from contact, mark run as responded
+        if not self.run.responded and msg.direction == INCOMING:
+            self.run.responded = True
+            self.run.save(update_fields=('responded',))
```
Comment (nicpottier) [inline temba/flows/models.py]: Do we consider a flow started from a trigger keyword as responded?

File: temba/flows/migrations/0044_flowrun_responded.py (hunk 2 above)
Comment (rowanseymour) [inline temba/flows/migrations/0044_flowrun_responded.py]: It's defaulting to false, then next migration only has to update runs where it's true.

File: temba/flows/models.py (hunk 3 above)
Comment (rowanseymour) [inline temba/flows/models.py]: The name `responded` perhaps doesn't work so well in that case, but I think the most useful question for this field to answer is "did a contact interact with this run" so yes.

Might be worth me adding a test for that case if you agree.

Comment (nicpottier) [inline temba/flows/models.py]: Ya I think having a test case to cement that contract makes sense.

File: temba/flows/migrations/0044_flowrun_responded.py (hunk 2 above)
Comment (nicpottier) [inline temba/flows/migrations/0044_flowrun_responded.py]: You can't add a non-nullable field to a table without locking it.. the process on big tables like flowrun needs to be.. add null field, populate it in a migration later, then switch it over.

**Comment by rowanseymour** [comment]
Reworked migrations to not lock runs table, so now it adds nullable field, populates it, then makes it not-nullable. Unfortunately migration is much slower now and takes ~20mins per million runs.

**Pull request merged**